### PR TITLE
Use reusable GitHub workflows for archiving and creating the XCFramework

### PR DESCRIPTION
**Pull request by philippzagar**
# Use template GitHub workflows for archiving and creating the XCFramework

## :recycle: Current situation & Problem
In https://github.com/StanfordBDHG/.github/pull/40, we introduced reusable GitHub Actions workflows for the archiving and creation of XCFrameworks. This enables us to avoid code duplication.

## :gear: Release Notes 
- GitHub Actions workflow utilizes shared reusable workflows.
- ResearchKit updated to latest version v2.2.12

## :books: Documentation
Provided in the workflow files

## :white_check_mark: Testing
Manually tested by creating an example release of ResearchKit.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [X] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philippzagar** [comment]
@PSchmiedmayer Just created a new release of ResearchKit with the most up-to-date code state from Apple. Works flawlessly! 🚀
Feel free to review the changes!

**Comment by philippzagar** [comment]
@PSchmiedmayer Just created a new release via the reusable GitHub Actions workflows, worked fine: https://github.com/StanfordBDHG/ResearchKit/releases/tag/2.2.19

Feel free to check the (minimal) code changes (most of the diff comes from upstream ResearchKit)  🚀

File: .github/workflows/build.yml
Comment (PSchmiedmayer) [inline .github/workflows/build.yml:34]: Is this default value needed here?

File: .github/workflows/build.yml
Comment (PSchmiedmayer) [inline .github/workflows/build.yml:24]: I would suggest that we don't even make this configurable as we want to ensure that releases are consistent and we know that we want release optimization for all releases here in this fork.

File: .github/workflows/build.yml
Comment (PSchmiedmayer) [inline .github/workflows/build.yml:8]: Nice, thank you!

File: .github/workflows/release.yml
Comment (PSchmiedmayer) [inline .github/workflows/release.yml:22]: Same comment as above 👍

File: .github/workflows/build.yml
Comment (PSchmiedmayer) [inline .github/workflows/build.yml:43]: As we do not set `ORK_FRAMEWORK_VERSION_NUMBER` and now use `VERSION_NUMBER` (https://github.com/StanfordBDHG/.github/pull/40/files) we should adjust this here in the Xcode configuration.

This env variable is used in various places and we should ensure that this is set properly. Best way would probably be a string replacement across the complete project.

File: ResearchKit.xcframework/Info.plist
Comment (PSchmiedmayer) [inline ResearchKit.xcframework/Info.plist:1]: Can you validate that all the builds still work and produce the same result by you locally checking out https://github.com/StanfordSpezi/SpeziQuestionnaire, setting the dependency on this feature branch, and then manually running the tests to ensure everything is building and not running into a runtime issue?

That would be amazing to ensure that we don't introduce any regressions with the new build setup and don't need a lot of follow up PRs 🚀

File: samples/ORKCatalog/ORKCatalog/Configuration/ORKCatalog-Shared.xcconfig
Comment (PSchmiedmayer) [inline samples/ORKCatalog/ORKCatalog/Configuration/ORKCatalog-Shared.xcconfig:15]: This might just be here in the example, but I think that we might need to check the diff to ensure that we do not introduce some regressions here. We set `CI` (https://github.com/StanfordBDHG/.github/blob/main/.github/workflows/archive.yml#L70) to true and not `USE_CI`. Might also affect some other build related scripts and elements in here, this is just the position where I first noticed it. This one might not even be relevant as this is the example but it would be great to check this in the different projects.

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for all the improvements. I had a few minor comments that I noted in the review, would be great if those are addressed before we merge the PR. 

Thanks for all the work here to make this reusable 🚀

No need for an other review from my side once these issues are addressed and you validated that the feature branch is working using an external package and tests 👍

File: .github/workflows/build.yml
Comment (philippzagar) [reply .github/workflows/build.yml:34]: I guess it was needed as the `configuration: default: 'Release'` is not applied for the `pull_request` trigger (as far as I understand the trigger condition). However, if we ditch the possibility to manually configure the build configuration all together, this is redundant 🚀

Comment (PSchmiedmayer) [reply .github/workflows/build.yml:34]: Oh, good point. But yes, I would prefer to just remove it to make it simper and avoid confusion in the future.

File: ResearchKit.xcframework/Info.plist
Comment (philippzagar) [reply ResearchKit.xcframework/Info.plist:1]: Yep, just tested that, compiles fine and also successfully passed the UITests 🚀

File: .github/workflows/build.yml
Comment (philippzagar) [reply .github/workflows/build.yml:43]: I'm a bit confused how these variables are used within ResearchKit:
https://github.com/StanfordBDHG/ResearchKit/blob/9530fadc9d09c0b783388c752b0dbe841974b6af/ResearchKit/Configuration/ResearchKit/ResearchKit-Shared.xcconfig#L30-L35
We could probably even delete some of the lines here (as we don't need it with our infrastructure)...

For now, I just did a string replacement of `ORK_FRAMEWORK_VERSION_NUMBER ` by `VERSION_NUMBER` for the entire project.

File: ResearchKit.xcframework/Info.plist
Comment (PSchmiedmayer) [reply ResearchKit.xcframework/Info.plist:1]: Perfect; thank you!

File: .github/workflows/build.yml
Comment (PSchmiedmayer) [reply .github/workflows/build.yml:43]: Yes, this is for their infrastructure, I think a string replacement for now is finde even though we don't use the more sophisticated mechanisms in the moment.

File: samples/ORKCatalog/ORKCatalog/Configuration/ORKCatalog-Shared.xcconfig
Comment (philippzagar) [reply samples/ORKCatalog/ORKCatalog/Configuration/ORKCatalog-Shared.xcconfig:15]: I'm not sure why Apple made this change within the file you mentioned, they are even inconsistent within the same file: https://github.com/StanfordBDHG/ResearchKit/blob/94d6a12b9c157a0f4716577a8a96ce5ddb16ccd7/samples/ORKCatalog/ORKCatalog/Configuration/ORKCatalog-Shared.xcconfig#L7

For now, I replaced the `USE_CI` with just `CI` (as we pass this flag via the `xcodebuild archive` cmd) but I'm not sure what else I should do here.

Comment (PSchmiedmayer) [reply samples/ORKCatalog/ORKCatalog/Configuration/ORKCatalog-Shared.xcconfig:15]: Sounds good, that should work! As far as I understand ORKCatalog is an own thing by itself put there for demonstration purposes and the overall setup is a bit unfortunate as the CI and the detailed setup is not open source and the project is not really maintained with a community aspect in mind. I think it is good to merge as it is.

**Pull request merged**
